electrum/plugins/bitbox02/bitbox02.py

```diff
@@ -0,0 +1,401 @@
+        address_keypath, is_mainnet = self.bip32_list_from_keypath_string(bip32_path)
+        if address_type == "p2wpkh" or address_type == "p2pkh" or address_type == "standard":
+            if is_mainnet:
+                coin_network = bitbox02.btc.BTC
+                script = bitbox02.btc.SCRIPT_P2WPKH
+            else:
+                coin_network = bitbox02.btc.TBTC
+                script = bitbox02.btc.SCRIPT_P2WPKH
+        elif address_type == "p2wpkh-p2sh":
+            if is_mainnet:
+                coin_network = bitbox02.btc.BTC
+                script = bitbox02.btc.SCRIPT_P2WPKH_P2SH
+            else:
+                coin_network = bitbox02.btc.TBTC
+                script = bitbox02.btc.SCRIPT_P2WPKH_P2SH
+        else:
+            raise Exception("invalid keypath address_type")
+
+        address = self.device.btc_pub(
+                keypath = address_keypath,
+                output_type = bitbox02.btc.BTCPubRequest.ADDRESS,  # pylint: disable=no-member
+                script_type = script,
+                coin = coin_network,
+            )
+        return address
+
+    def sign_transaction(self, keystore, tx):
+        if tx.is_complete():
+            return
+
+        inputhasharray = []
+        hasharray = []
+        pubkeyarray = []
+        network = bitbox02.btc.BTC
+        if constants.net.TESTNET:
+            network = bitbox02.btc.TBTC
+
+        tx_script_type = bitbox02.btc.SCRIPT_P2WPKH
+
+        # Build BTCInputType list
+        inputs = []
+        for txin in tx.inputs():
+            if txin.is_coinbase():
+                self.give_error("Coinbase transactions are not supported")
+            _, full_path = keystore.find_my_pubkey_in_txinout(txin)
+
+            inputs.append({
+                "prev_out_hash": txin.prevout.txid,
+                "prev_out_index": txin.prevout.out_idx,
+                "prev_out_value": txin.value_sats(),
+                "sequence": 0xffffffff,
+                "keypath": full_path,
+            })
+
+            if full_path[0] == 49 + HARDENED:
+                tx_script_type = bitbox02.btc.SCRIPT_P2WPKH_P2SH
+
+        # Build BTCOutputType list
+        outputs = []
+        for txout in tx.outputs():
+            assert txout.address
+            # check for change
+            if txout.is_change:
+                _, full_path = keystore.find_my_pubkey_in_txinout(txout)
+
+                outputs.append(
+                    bitbox02.BTCOutputInternal(
+                        keypath = full_path,
+                        value = txout.value,
+                    )
+                )
+            else:
+                addrtype, pubkey_hash = bitcoin.address_to_hash(txout.address)
+                if addrtype == constants.net.ADDRTYPE_P2PKH:
+                    output_type = 1
+                elif addrtype == constants.net.ADDRTYPE_P2SH:
+                    output_type = 2
+                else:
```

Comment:
... then you can also do an explicit check here, and the `else:` part can raise an Exception.